hibernate-core/src/test/java/org/hibernate/jpa/test/callbacks/PreUpdateModificationsLazyLoadingTest.java

```diff
@@ -0,0 +1,90 @@
+package org.hibernate.jpa.test.callbacks;
+
+import org.hibernate.cfg.AvailableSettings;
+import org.hibernate.jpa.test.BaseEntityManagerFunctionalTestCase;
+import org.hibernate.testing.TestForIssue;
+import org.hibernate.testing.bytecode.enhancement.BytecodeEnhancerRunner;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import javax.persistence.*;
+import java.nio.ByteBuffer;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+
+import static org.junit.Assert.*;
+
+@TestForIssue( jiraKey = "HHH-12718" )
+@RunWith(BytecodeEnhancerRunner.class)
+public class PreUpdateModificationsLazyLoadingTest extends BaseEntityManagerFunctionalTestCase {
+
+    @Test
+    public void testPreUpdateModifications() {
+        EntityManager em = getOrCreateEntityManager();
+        em.getTransaction().begin();
+        Person p = new Person();
+        em.persist(p);
+        em.getTransaction().commit();
+
+        em.clear();
+
+        em.getTransaction().begin();
+        p = em.find(Person.class, p.id);
+        assertNotNull(p);
+        assertNotNull(p.createdAt);
+        assertNull(p.lastUpdatedAt);
+
+        p.name = "Changed Name";
+        em.getTransaction().commit();
+
+        em.clear();
+
+        p = em.find(Person.class, p.id);
+        // This last assertion fails
+        assertNotNull(p.lastUpdatedAt);
+        em.close();
+    }
+
+    @Override
+    protected Class<?>[] getAnnotatedClasses() {
+        return new Class<?>[] { Person.class };
+    }
+
+    @Override
+    protected void addConfigOptions(Map options) {
+        options.put(AvailableSettings.CLASSLOADERS, getClass().getClassLoader() );
+        options.put(AvailableSettings.ENHANCER_ENABLE_LAZY_INITIALIZATION, "true");
+        options.put(AvailableSettings.ENHANCER_ENABLE_DIRTY_TRACKING, "true");
+    }
+
+    @Entity(name = "Person")
+    private static class Person {
+        @Id
+        @GeneratedValue
+        private int id;
+
+        private String name;
+
+        private Instant createdAt;
+
+        private Instant lastUpdatedAt;
+
+        @ElementCollection
+        private List<String> tags;
+
+        @Lob
+        @Basic(fetch = FetchType.LAZY)
+        private ByteBuffer image;
+
+        @PrePersist
+        void beforeCreate() {
+            createdAt = Instant.now();
```

Comment:
Aren't you supposed to use a `setCreatedAt` method here? I think dirty tracking enhancement is not triggered with field access (and the dirty tracking optimisation would be useless if the fields where dirty tracked anyway regardless of the optimisation being enabled).

Reply:
Hi Jan Willem, you're right. I'll update the test case to use setters.
However I think it would make a difference with regards to the underlying bug that I wanted to isolate here. The direct field assignment p.name = "Changed Name" is recognized as dirty field. The interesting behavior is that as soon as you remove one of the lazily loaded fields the test case goes through.

Reply:
Thanks for editing the test case! I see it still fails indeed, interesting bug.